preview implementation of the Laminas Integration RFC

Jurj-Bogdan opened the pull request:
|    Q          |   A
|-------------- | ------
| New Feature   | yes
| RFC           | yes


### Description

This PR is an initial implementation for the [Laminas ecosystem RFC](https://github.com/laminas/getlaminas.org/issues/199). There are still features I think should be implemented (i'll list a few at the end), but hopefully there's enough for a preview to help continue the conversation started in the RFC.

### Current functionality

The `laminas ecosystem:create-db` and `laminas ecosystem:seed-db` commands are used to generate and update the list found on '/ecosystem' by using the user-provided data from `ecosystem-packages.json`.

As the consensus from the RFC is that packages must be installable via composer, the Packagist url is mandatory and used to make a single request per package, from which the displayed data is extracted. 
If provided, the 'homepage' and 'githubUrl' data will be used to overwrite the data coming from Packagist, and the 'categories' are used as extra filtering options in the package listing.

The `create-db` command currently works similar to the blog version, regenerating the whole database file after each build, although i would change this to make it more future proof. I've thought of the  `seed-db` command as a cron used to select packages not updated in X hours and update some of the relevant data - would something real-time (or closer to) be desired instead? 

As for the listing page ('/ecosystem') I've left a low page size to make it easier to test the pagination with the few default packages. For now there's searching by package name and filtering by "tags" (keywords taken from packagist) and "categories" defined by the user in the json file.

#### what next?
- i'm aware of the failing psalm check, but seeing as this PR is subject to change, I'll make sure to clean that up once i have a better understanding of what is required from this feature
- maybe adding a github action to validate the `ecosystem-packages.json` file, to make it easier for the reviewer ?
- if the current approach is kept, i think changing the database creation command to only add the newest entries if the db already exists would be a good change
- design changes
- as both the [blog search](https://getlaminas.org/blog/search?q=laminas) and [maintenance status](https://getlaminas.org/maintenance-status) features provide json responses, this could be added to the ecosystem package list as well if desired
- expanding the instructions for adding a new package once the desired rules and approach is agreed on

froschdesign commented:
@Jurj-Bogdan 
What was the link to the preview?

froschdesign commented on data/ecosystem/ecosystem-packages.json:
```diff
@@ -0,0 +1,42 @@
+  {
+    "packagistUrl": "https://packagist.org/packages/netglue/laminas-messenger",
+    "githubUrl":  "https://github.com/netglue/laminas-messenger",
+    "categories": ["laminas", "messenger"],
```
"laminas" does not make sense here, because we need to differentiate between integration in Mezzio-based and/or laminas-mvc-based applications. This means that the package can be used as a:

- module in a laminas-mvc-based application
- and/or via a config provider in a Mezzio-based application

Maybe the categories are not suitable for this and a separate entry or entries are required.

Jurj-Bogdan replied:
At the moment the "categories" key is only used for filtering in the interface, so I added a few here and there so they'll be visible in the preview - once a new role is found for them, I can upgrade the functionality to fit the request of course.

Even if the current approach of "user defined keywords" is kept, some other guidelines could be defined for them if wanted.

arhimede replied:
we need to brainstorm a bit about those possible categories.

arhimede commented:
PREVIEW HERE:
https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/ecosystem/

Jurj-Bogdan commented:
A few details and questions about the latest commit:

1. The packagist api doesn't give access to the repository's social preview, so I've added a separate call to github's graphql api to only bring the image URL. A few issues here:
    - if a repository has no custom social preview, an image will be generated for it which will double some already available info (name, number of stars, example below), but I can't remove said information as I don't have a way to know what info is displayed in the preview (i.e. for a custom preview such as dotkernel api's that doesn't provide a name etc.)
        - to note is that the packagist api provides access to the repository owner's avatar, so if wanted this avatar could be used instead, so the separate github api call can be removed - however this means custom social previews would be ignored, it's your call
        - another option would be to add an optional "image" key to the json file, where the user would manually submit a URL to be displayed
![image](https://github.com/user-attachments/assets/06688ad8-4eec-4b85-a89f-5ef9ea1b4b98)
![image](https://github.com/user-attachments/assets/85979880-363e-476b-bd63-6b54b7beeac0)

2. Is this version of getting data from the api ("create-db" command run on build + "seed-db" command used as a cron for data refreshing every X hours) alright with you or would another way be desired? If it's OK i'd like to (and should) go ahead and clean up the code alongside any future changes.
    - In the initial commit, the "create ecosystem db" command worked similar to the blog one by deleting the existing db file and regenerating in on each build which will be problematic with large numbers of packages added. I've changed this so the existing database is kept and packages existing in the db are skipped when getting data from the api. This way will also ensure different "created/updated" dates for the cron to update packages in smaller batches.

3. about the data saved and displayed for each package:
    - i've kept the keywords taken from the package's `composer.json` and the user-defined keywords from the `ecosystem-packages.json` and am basically using them as filters, however they can clutter up the card in large numbers - do you think any of them should be removed/limited in number?
    - the Composer package type is now pulled from the API and displayed as it feels relevant, would filtering by type be a good addition?
    - similar to the above, i've added a user-defined "category" with the "skeleton", "integration" (or "wrapper" as mentioned in the original issue)  and "tool" values available, which again seem to be a useful information to display and maybe to filter based on the intended use case - should I go ahead with this feature?

arhimede commented:
Preview updated 
https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/ecosystem/

arhimede commented:
now the social preview is not looking good 
![image](https://github.com/user-attachments/assets/3bc284af-6095-4791-90bc-a5d4aa9d5d87)

We will try a variant to use as image or "social preview" only the github organisation logo .

gsteel commented:
My 2 cents:

1. I think that using the GH social preview is probably the best plan - user avatars are a bit hit and miss and also less relevant - For example, the avatar for all the laminas packages is @weierophinney gravatar:
   ![Mwops avatar](https://www.gravatar.com/avatar/ec1ea075e2cde62a611373c2b3cc1c70?s=48&d=identicon)
  Can we get away with no image at all as an alternative or does the layout get too boring? Also, what about libs hosted on GitLab/Codeberg etc?
2. Retaining the DB seems like a sane plan 👍 
3. Usage of the composer package type is really worth-while - I don't think the concept of "type" should be extended beyond the ones composer offers, that said "category" is helpful, but I believe these should be small in number, predefined and packages should only appear in one category. The main problem I see here is: lots of "integrations" and a handful of "tools" and "skeletons" at which point category starts feeling less useful.
  I think that the main filtering should be done via user-supplied tags that are audited by us prior to merge, rather than relying on GH topics or composer keywords - this will be the main tool that will allow us to curate collections so that they are meaningful for users.

gsteel commented:
Also, rather than "See on GitHub", perhaps just "Source" with the icon for the relevant source code hosting platform?

weierophinney commented:
How are you/will you determine packages to include? I ask, because, for instance, the phly-simple-page package on there likely should not be (I've not updated it in 3 years, and am in fact planning on archiving all my ZF2 packages (even if I did update them for Laminas at some point).

arhimede commented:
> How are you/will you determine packages to include? I ask, because, for instance, the phly-simple-page package on there likely should not be (I've not updated it in 3 years, and am in fact planning on archiving all my ZF2 packages (even if I did update them for Laminas at some point).

All packages we have listed now are only test data.
In order to decide the layout and the logic of that page.
Will be subject of a future TSC meeting to decide the rules to include and remove a package from that Ecosystem page

froschdesign commented:
1. You're going to hate me for this, but the term "ecosystem" is wrong here, because a few modules, packages or whatever you want to call them cannot be the ecosystem of Laminas. The listed packages/modules are _a part_ of the ecosystem, but not the entire system.
And as an end user, I would also not search for or suspect extensions and libraries from third parties under this name.
3. The amount of information can be greatly reduced, because licences, downloads and various links are too much. A link to the repository and - if available - to the documentation/homepage is sufficient. (Is the link to Packagist really helpful for a user? After all, it only redirects to the repository.)
4. If we also want to have packages/modules hosted on Gitlab or other platforms, then we need to rethink the information gathering process.
5. I would also like to point out again that we need a clear distinction between what can be used in an MVC application and what can be used in Mezzio.
For example, [LmcUser](https://lm-commons.github.io/LmcUser) must be listed and it must be clear to the end user that this only works in a laminas-mvc based application.

froschdesign commented:
A consistent layout would also be desirable, and the unnecessary spacing due to paragraphs is not necessary. 

![preview-packages](https://github.com/user-attachments/assets/f128697c-c74b-4133-8fbc-0b9239da21f8)

The description text at the end, then the problems are reduced. Example: https://getgrav.org/downloads/plugins

arhimede commented:
> 1. You're going to hate me for this, but the term "ecosystem" is wrong here, because a few modules, packages or whatever you want to call them cannot be the ecosystem of Laminas. The listed packages/modules are _a part_ of the ecosystem, but not the entire system.
>        And as an end user, I would also not search for or suspect extensions and libraries from third parties under this name.

We can call it "directory" instead , even the word "ecosystem" is more catchy :-) 
This will be a subject of a vote in TSC.

> 2. The amount of information can be greatly reduced, because licences, downloads and various links are too much. A link to the repository and - if available - to the documentation/homepage is sufficient. (Is the link to Packagist really helpful for a user? 
> After all, it only redirects to the repository.)

 Agree, now it is too much information.
@Jurj-Bogdan 
One nice directory is the one for [React Directory](https://reactnative.directory/)

>3. If we also want to have packages/modules hosted on Gitlab or other platforms, then we need to rethink the information gathering process.

The information will be gathered *only* from Packagist, as I presume a mandatory condition for a package to be lsited here will be to be installable using Composer.

>4. I would also like to point out again that we need a clear distinction between what can be used in an MVC application and what can be used in Mezzio.
> For example, [LmcUser](https://lm-commons.github.io/LmcUser) must be listed and it must be clear to the end user that this only works in a laminas-mvc based application.

@Jurj-Bogdan we need a Flag here, separate from "category". 
MVC, Middleware, Mezzio, will see which one will be the default flags

froschdesign commented:
By the way, this is boring: https://symfony.com/bundles 😉

Jurj-Bogdan commented:
I have started tweaking the info such as the keywords from composer being taken out, but keeping the user defined ones (probably going to limit those to 5? as well). Will also provide a sort of flag as @arhimede mentioned, with preset values which should be used as filters as well (thinking of adding a navbar around the search input for these "preset filters" - category, type and the new "flag")

Thanks @froschdesign for the example, i'll work on a version with the social preview working similar to getgrav.org's images; that is unless @gsteel 's question of removing images altogether is preferred, of course.

froschdesign commented:
Another example: https://astro.build/integrations/
Also interesting what information is required: https://docs.astro.build/en/reference/publish-to-npm/#integrations-library

gsteel commented:
Adding relevant keywords to your composer package is a great idea - I wonder if the packagist api can handle that? We'd need a way of excluding packages too though.

froschdesign commented:
> We'd need a way of excluding packages too though.

The packages must be added manually, not automatically! Or have I missed something?

gsteel commented:
I was just ruminating on being able to automate that - it's probably best to stick to the plan and add them manually as originally specified 👍

froschdesign commented:
Adding must be done proactively and I'm not worried about that, because some people are already waiting for an official listing.
The next step would be a new series of blog articles under the name "Package/Module of the Month", or something similar.

arhimede commented:
updated preview 
https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/ecosystem/

Jurj-Bogdan commented:
an overview for the latest commits:

1. kept the social preview as suggested by @gsteel, with a small on-hover effect to display more information, but I have a few questions related to it.
   The image has to be relatively large for the small details to be somewhat legible at least (especially for the auto-generated social previews), but to me it feels like the cards get a bit too large, so:
    - would it be better for the social preview to hide completely on-hover as opposed to the "cut off" that is up now? I could also hide more information by default, with it only popping up on hover (hide the source, download/star numbers and the description) resulting in smaller cards
    - the "type", "usage" and "category" line is also weird, the info is relevant and simply replacing the words with icons could work, but might feel confusing about which is which i think - any preference here?
    - i feel there should be some limits to the user-defined keywords, what do you think? A limited number, maybe even a character limit per keyword?
    - one issue right now is related to packages hosted on anything other than github - for these there obviously won't be any social preview to grab with the current command, so I provided some default images (by category: [skeleton](https://icons.getbootstrap.com/icons/stack/), [integration](https://icons.getbootstrap.com/icons/plug/) & [tool](https://icons.getbootstrap.com/icons/wrench/))
        - these will be if there are any errors when getting the social preview, such as for `phly/phly-simple-page`, which has another name on github - I'll fix the way previews are brought to avoid this exact issue, but i think it can be useful for others

2. added filters by type, usage and category:
    - "type" - values are taken from packagist: "library", "project", "metapackage", "composer-plugin". Should all these types be allowed?
    - "usage" - to provide a distinction between mvc / mezzio application, as suggested by @froschdesign and @arhimede 
    - "category" - can be one of "skeleton", "integration" and "tool", the 
Filters can be removed individually by clicking each again in the dropdown. The "clear" filters button will not touch the search/pagination, but only remove these 3.

3. added a flag for abandoned packages in case it's decided these would be worth displaying - let me know if they aren't desired so I can skip them altogether if any gets submitted and approved. If so, i'll have to extend the "refresh data" cron to remove any package that gets abandoned after being accepted as well

weierophinney commented:
A few notes:

- I wasn't sure at first about the hover effect, until I found one where the content got truncated and wasn't fully visible until the hover. As such, I think it works.
- If a package is marked abandoned, I don't think we should display it (mezzio-aurarouter is the example here).
- It would be nice to be able to clear individual filters, instead of needing to clear _all_ filters.
- Will there be pagination at some point, or will we always show everything? What will the performance be like if we show everything?
- We need some sort of explanation for the "Category" filtering; I'm not sure what a tool is in this context.

Jurj-Bogdan commented:
- individual filters can be cleared by clicking them again in the dropdown - maybe I should add an "X" so it stands out more.
- there's a 15 packages per page limit right now and not enough packages, but yes, there is pagination available.
- the categories aren't really intuitive, i agree, but once agreed on a set of categories i'll make sure to explain every term in an updated  `ADD_ECOSYSTEM_PACKAGE.md` file. As it stands right now, the categories are supposed to be something like this:
   - "skeleton" - package that allows building an entire application upon
   - "integration" - package that extends an existing package's functionality
   - "tool" -  "standalone" package that provides its own functionality;
   -  i should add i've set this category randomly across the existing package list so there are a few available for filtering in the preview

arhimede commented:
> > * but once agreed on a set of categories i'll make sure to explain every term in an updated  `ADD_ECOSYSTEM_PACKAGE.md` file.
> 
> But this does not help the user / visitor of the website. The filter must be self-explanatory on the website. I assume that this is what @weierophinney means.

probably the word "Tool" is not the best here. We may find another one. 
I want to differentiate between those 2 types of libraries:
- https://packagist.org/packages/akrabat/ip-address-middleware which is a standalone library, which are doing one thing, and does not have external dependecies (other then PSR's)
- https://github.com/netglue/laminas-messenger which is an "Integration" , or "Wrapper" over Symfony messenger library

arhimede commented:
Preview updated 
https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/ecosystem/

froschdesign commented on bootstrap/js/_ecosystem.js:
```diff
@@ -0,0 +1,114 @@
+'use strict';
+
+$(document).ready(function () {
```
Vanilla JavaScript is good in 2024 and it should be dropped when we move to Bootstrap 5.

Jurj-Bogdan replied:
Fair enough, I saw jquery was already added as a dependency and went with it, but I'll go ahead and refactor

Jurj-Bogdan commented:
Added an extra GH workflow to make sure the .json file being edited by users is not empty and is valid JSON, as well as making sure all the required keys are defined (should I also extend it so the values are checked as well?).

Let me know if another way to handle this would be preferred, and also of any other changes to the design and/or logic as well, of course.

froschdesign commented on src/Ecosystem/templates/list.phtml:
```diff
@@ -0,0 +1,187 @@
+                    </div>
+
+                    <div class="btn-group">
+                        <button id="clear-filters-button" type="button" class="btn btn-outline-danger">Clear filters <i class="bi bi-x-circle"></i></button>
```
The button to clear the filter should only be visible if the user has set a filter and should also reset the search. Otherwise I find the handling confusing.

Jurj-Bogdan commented:
Replaced jQuery with vanilla JS in the custom script, and the preview has been updated: https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/ecosystem/

I'd would need some input for a few pretty major issues:

1. The "package must be available via Composer / Packagist" rule seems set in stone at the moment, however since Packagist allows packages hosted on GitLab, Bitbucket etc., any calls to GitHub's API for the social preview will fail. For now that will result in default images being shown
    - should I look into getting data for such repositories?
    - if not, maybe some other type of default images should be shown? Currently they're simple bootstrap icons based on category: [skeleton](https://icons.getbootstrap.com/icons/stack/), [integration](https://icons.getbootstrap.com/icons/plug/) & [tool](https://icons.getbootstrap.com/icons/wrench/)

2. do you feel the the implementation itself (commands used for data gathering, storing packages in the sqlite db etc.) requires major changes? Because if not, I'll go ahead and clean up the code and fixing the Psalm errors
    - also if this implementation stays, do you think the commands should use a package to make the API calls instead of the current native cURL?

3. somewhat related to the previous question, once the implementation is "set" should i go ahead and add tests?

weierophinney commented:
> Replaced jQuery with vanilla JS in the custom script, and the preview has been updated: https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/ecosystem/
> 
> I'd would need some input for a few pretty major issues:
> 
> 1. The "package must be available via Composer / Packagist" rule seems set in stone at the moment, however since Packagist allows packages hosted on GitLab, Bitbucket etc., any calls to GitHub's API for the social preview will fail. For now that will result in default images being shown
>    
>    * should I look into getting data for such repositories?
>    * if not, maybe some other type of default images should be shown? Currently they're simple bootstrap icons based on category: [skeleton](https://icons.getbootstrap.com/icons/stack/), [integration](https://icons.getbootstrap.com/icons/plug/) & [tool](https://icons.getbootstrap.com/icons/wrench/)

Yes - exactly this last approach. If we can't look up an image, use one based on the category.

> 2. do you feel the the implementation itself (commands used for data gathering, storing packages in the sqlite db etc.) requires major changes? Because if not, I'll go ahead and clean up the code and fixing the Psalm errors

Assuming you address the default image solution, yes.

>    * also if this implementation stays, do you think the commands should use a package to make the API calls instead of the current native cURL?

Since we can control the environment where the command runs, cURL is fine.

> 3. somewhat related to the previous question, once the implementation is "set" should i go ahead and add tests?

Yes, please.

Jurj-Bogdan commented:
a rundown of the latest update:

1. added package removal by deleting the entry from the `ecosystem-packages.json` file, so anytime a package is removed from the file it will also be removed from the database.
    - to note is that I haven't automated removal of packages in case they become abandoned after being added; abandoned packages are treated as follows:
    - any package that is abandoned before being added will be ignored by default (as "mezzio/mezzio-aurarouter" is right now in the file, kept there for testing purposes)
    - any package that is abandoned after being successfully added will be marked as abandoned by the cron and will display a warning on the listing page, but removal will have to be manual via commit
    - let me know if another option is preferable

2. psalm fixes
3. added an initial batch of unit tests
    - adding functional tests for the commands that write files will require [mikey179/vfsstream](https://packagist.org/packages/mikey179/vfsstream) or an alternative to mock creating/writing the db, would that be an issue?

PowerKiKi commented:
I would remove images entirely, because they duplicate info, such as the most important one: the title. They are obviously sub-optimal for SEO. And they are hard to read because text is too small and contrast too poor, especially on a 770px wide screen.

Then we can get rid of the mouse over animation too, so that the `card-text` is entirely readable without having to move the mouse around (and avoid the "jumpy" effect it introduces).

If you want to keep some sort of image, then I would use the user/organization profile picture, which is guaranteed to have a more uniform layout, and provide a strong visual indicator of the library author(s).

In the preview `card-updated` are all the same. If it will always be that way, then we don't need to repeat it everywhere. One mention somewhere on the page (header or footer ?) should be enough.

Similarly the "homepage" link is very often the same as the source link. In that case it brings no value, and should not be shown. In fact, I'd even suggest to never show the homepage link, because we already have two links (packagist and source) for basically the same thing.

froschdesign commented:
> Similarly the "homepage" link is very often the same as the source link. In that case it brings no value, and should not be shown. In fact, I'd even suggest to never show the homepage link, because we already have two links (packagist and source) for basically the same thing.

If the target is the same, then it really makes no sense.

---

@arhimede @Jurj-Bogdan 
Individual notes, corrections and tasks listed here should be kept somewhere. Possibly as [sub-issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/adding-sub-issues).

arhimede commented:
> @arhimede @Jurj-Bogdan Individual notes, corrections and tasks listed here should be kept somewhere. Possibly as [sub-issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/adding-sub-issues).

Yup, we are about to update this PR tomorrow with some modifications and corrections
As i want to release this feature before jumping on implementing the new design

Jurj-Bogdan commented:
@arhimede please push to any free preview branch when possible

Quick rundown on the latest commit:
1. As the github social preview seems to be a main issue in the card design, I've replaced it altogether with the repository owner's avatar (as opposed to a previous version which displayed maintainer avatar)
    - while user avatars were available directly from the packagist api request, repo owner avatars are only available via a separate call to the github api, but this was the case for the social preview as well, so hopefully that's still not a problem
2. instead of saving the url to the image as was the case for social previews, the avatars are now saved locally to limit external calls to load images
3. removed the "pop out" effect and made the cards somewhat smaller
    - somewhat related to this, should a limit be set for the displayed description? As this is taken directly from packagist, nothing is stopping a "wall of text" being set as a package's description, which will then be abruptly cut off in the card
    - should i impose a character limit on the displayed text only (maybe extend on click?) or make the description user defined as well, with a new key under the `ecosystem-packages.json` file? Obviously open to suggestions

Also a few details related to @PowerKiKi 's comment:
- the current `card-updated` sections are the same as this initial batch of packages was added at the same time, with any subsequent packages added having their own update times (the update cron runs every 12h, collecting only the packages not updated for longer)
- the `homepage` button is configurable via the json file, allowing users to submit a website of their choosing (I've set `dot-errorhandler`  to point to dotkernel.com for example); if nothing is set by the user, the `homepage` is set from packagist if any is available, with the button skipped if none. If you want I can either remove it completely or only set it if found in the `ecosystem-packages.json` file

PowerKiKi commented:
I like this version better :smile: 

For the filters I would suggest to keep them closer together and to show explicitly that nothing is filtered yet. Something similar to this:

![image](https://github.com/user-attachments/assets/f4f0f41d-8b9a-4b65-b27a-1d55e4303218)

When tags are selected, the mouse pointer does not show it is clickable, it stays as either the text selection pointer or the default pointer. Also maybe a sight hover effect would help to understand it is interactive.

![image](https://github.com/user-attachments/assets/212207f1-dccd-4c98-8c46-7d94fb31b3ec)


On a small screen the profile picture aspect ratio get distorted. The example here is a bit of an extremely small screen, but it could also easily happen if the lib name happens to be long-ish:

![image](https://github.com/user-attachments/assets/8218d3a9-5aff-480f-8299-7921b47a0d39)

Thank for the explanation about `card-updated`. I would maybe consider adding a tooltip to explain what it is. And/or maybe consider use a lib like timeago to show something like "info updated 6 hours ago" ?

PowerKiKi commented:
In https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2025-03-03-TSC-Meeting.md#criteria-for-including-in-ecosystem-page the JSON format is described. What about formalizing all of that with a JSON schema that lives next to, and is referenced by, `ecosystem-packages.json` ? So that future contributors don't need to guess the JSON format, and instead can rely on their editor to suggest possibles values and highlight invalid things ?

I guess it would save a lot of back and forth for newcomers that attempt to add a new package.

arhimede commented:
Preview is here 
https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/ecosystem/

@laminas/technical-steering-committee let's discuss this too at TSC meeting July

froschdesign commented:
I think everything related to MVC should be removed.

Jurj-Bogdan commented:
Removed the "usage" filter completely, as it was either "mezzio" or "MVC", so it was useless without "mvc"... unless of course other options are to be added, but in that case it should be simple enough to reimplement

arhimede commented:
> I find the distinction between type and category confusing. Do we really need both selections or can this be reduced to one?

`type` is comming from packagist, so it is a propriety set in composer.json for each package 

`category` is something that was discussed previously , can be there or can be remove
Anyway, thi ecosystem page I think is a major part of the getlaminas.org, and we may want to have a box in home page with latest packages, or such, in order to emphasys more that ecosystem directory

PowerKiKi commented:
I guess the title of the page could be:

```diff
- Laminas Ecosystem
+ Laminas third party packages
```

since that is how the page is described.

I keep tring to click on the non-interactive part of "Type".

<img width="130" height="93" alt="image" src="https://github.com/user-attachments/assets/493e3efc-d789-4e4f-bbc9-229a719dffb5" />

Can't we have an entirely interactive dropdown button whose label concatenate "Type" and its value ? So the button is big enough to click on (it wouldn't if "Type" was totally outside of the button). And all of it is clickable. So when I want to change the type, I can click on "Type" or on "All".

<img width="117" height="67" alt="image" src="https://github.com/user-attachments/assets/7c9c3029-ede4-455b-a6f3-563ebdb98559" />

The Clear button has a nice hover effect indicating interactivity. But keyword buttons do not, in both places. I they they should have a hover effect too.

<img width="752" height="282" alt="image" src="https://github.com/user-attachments/assets/d02ca665-4aeb-40e3-895c-c52b1aee3ed7" />

Also the page lack information on how to add packages to that list. What is the process to maintain this list ? the process should at least be linked from this page.

Jurj-Bogdan commented:
Thanks for the feedback!
Just made the suggested filter changes, and added a small button that links to the upcoming [ADD_ECOSYSTEM_PACKAGE](https://github.com/laminas/getlaminas.org/blob/6d8a00725d0390e864f44c4ea3beca7c846da2b6/ADD_ECOSYSTEM_PACKAGE.md) - however this file is definitely subject to change, since as far as i know there is no final structure for the entries yet.

arhimede commented:
https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/ecosystem/

category drop-down was not removed yet,

samsonasik commented:
is there menu link to go to the page?

froschdesign commented:
@PowerKiKi 
> I guess the title of the page could be:
> 
> ```diff
> - Laminas Ecosystem
> + Laminas third party packages
> ```
> 
> since that is how the page is described.

Much better! 👍🏻  


> That's _very_ subjective, so do whatever you will with this remark, but I am not super fan of the pink-violet-gradient gradients.

Please create a separate issue report for this otherwise we will never be finished here.
Thanks in advance! 👍🏻

froschdesign commented:
@arhimede 
> I wonder if we can go live with this ?

The name "ecosystem" **must not** be visible from the _outside_. Change the URL, the "Adding your entry" file, the Readme file, etc.
There is also a conflict with the template file for the footer.

arhimede commented:
> @arhimede
> 
> > I wonder if we can go live with this ?
> 
> The name "ecosystem" **must not** be visible from the _outside_. Change the URL, the "Adding your entry" file, the Readme file, etc. There is also a conflict with the template file for the footer.

Laravel use the term `ecosystem` https://ecosystem.laravel.io/
React use the term `directory` https://reactnative.directory/

I am personally in favor of the term `ecosystem` , mostly because here we want to collect in time every 3rd party `stuff` that can be used in a Laminas project.
Not only libraries and packages, but also utilities related to server deployment, CI/CD, testing , etc . 
Also, due to Laravel popularity, developers are familiar with the term `ecosystem`.

We can use any of the below terms:
- community
- 3rd-parties
- directory
- hub
- vault
- suite
- galaxy
- atlas
- center
- gateway

Maybe should i open a Poll in slack about that ?

froschdesign commented:
> I am personally in favor of the term `ecosystem` , mostly because here we want to collect in time every 3rd party `stuff` that can be used in a Laminas project.

Others use the term much more broadly: https://vuejs.org

> We can use any of the below terms:

"Integrations" - https://astro.build/integrations/

arhimede commented:
new preview, using `integrations`
https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/integrations/

@Jurj-Bogdan please add a legend text block below title, to explan what is what

froschdesign commented:
A new entry should be added to the main menu, after "Docs".
And the headline should be "Integrations" not "Laminas Integrations".

arhimede commented:
preview updated

https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/

froschdesign commented:
@arhimede @Jurj-Bogdan 
> @froschdesign can we go live with it ?

Unfortunately not because we have to check the current listed packages:

- roave/psr-container-doctrine is not a middleware
- I like league/commonmark but where is the integration for Laminas?
- ramsey/uuid-doctrine is a integration for Doctrine not for Laminas
- ramsey/uuid can be used anywhere, so it is fine even if it is not a direct integration (can be discussed)

For example, the following packages are missing:

- [inertia-psr15](https://github.com/cherifGsoul/inertia-psr15)
- [zf-dependency-injection](https://github.com/reinfi/zf-dependency-injection)
- [mezzio-test](https://github.com/trinet-at/mezzio-test)
- [TwbsHelper](https://github.com/neilime/twbs-helper-module)
- [TacticianModule](https://github.com/mikemix/TacticianModule)

Jurj-Bogdan commented:
Oh, the current packages are there only for testing on the preview page, picked randomly from a list in the initial RFC and related packages from their dependencies, even including intentionally invalid packages.

For now I'll go ahead and remove invalid packages and add the listed ones

arhimede commented:
preview updated
https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/integrations/

PowerKiKi commented:
If you know how many pages there are, which is most likely the case, it would be nice to have proper pagination links, such as page 1,2,3... and also number of records found. Something with similar features as the picture below. Though I suppose it could come in a later PR.

<img width="491" height="54" alt="image" src="https://github.com/user-attachments/assets/5a75fd58-051d-4254-9503-17bf58fa0da9" />

You're gonna hate me for this, but.... The page is now called "Integrations". And on that page we can choose packages that are of category "integration", and other packages that are of category "tool". First of all, the difference between "tool" and "integration" is not immediately obvious to me, especially because "tool" is a very generic term. But the situation is even more confusing when we are on a page called "Integrations" but showing things that are specifically not "integration". Does it not feel a bit contradictory ?

froschdesign commented:
@PowerKiKi 
> The page is now called "Integrations". And on that page we can choose packages that are of category "integration", and other packages that are of category "tool". First of all, the difference between "tool" and "integration" is not immediately obvious to me, especially because "tool" is a very generic term. But the situation is even more confusing when we are on a page called "Integrations" but showing things that are specifically not "integration". Does it not feel a bit contradictory ?

You are absolutely right, it irritates me too!
There are types, categories and keywords, which may be too much.

PowerKiKi commented:
On another topic, tags should be searchable. If I search for "guid", I expect to find ramsey/uuid, because is has the tag "guid". Same goes for the description, I search for "PSR-15" I expect to find akrabat/ip-address-middleware.

And if a search really has no result, it would be a better UX to show a "no packages found for your search" kind of message.

arhimede commented:
> Some more packages:
> 
>     * [mezzio-navigation](https://packagist.org/packages/mimmi20/mezzio-navigation)
> 
>     * [laminasviewrenderer-vite-url](https://packagist.org/packages/mimmi20/laminasviewrenderer-vite-url)
> 
>     * [laminasviewrenderer-bootstrap-navigation](https://packagist.org/packages/mimmi20/laminasviewrenderer-bootstrap-navigation)
> 
>     * [laminasviewrenderer-js-log](https://packagist.org/packages/mimmi20/laminasviewrenderer-js-log)
> 
>     * [laminasviewrenderer-revision](https://packagist.org/packages/mimmi20/laminasviewrenderer-revision)
> 
>     * [laminasviewrenderer-bootstrap-form](https://packagist.org/packages/mimmi20/laminasviewrenderer-bootstrap-form)
> 
>     * [mezzio-setlocale-middleware](https://packagist.org/packages/mimmi20/mezzio-setlocale-middleware)
> 
>     * [laminas-form-element-paragraph](https://packagist.org/packages/mimmi20/laminas-form-element-paragraph)
> 
> 
> @mimmi20 Which of your packages should also be listed?
> 
> (All readme files missing a description what each packages provides. The user have to look at packagist.org or in the Composer configuration file!)

Let's wait for this ones until the package will have a description

PowerKiKi commented on data/integration/integration-packages.json:
```diff
@@ -0,0 +1,172 @@
+    "keywords": ["mezzio", "middleware", "language"],
+    "homepage": ""
+  }
+]
```
I'd like to add one of my package too. It is a middleware to support file upload in GraphQL via [webonyx/graphql-php](https://github.com/webonyx/graphql-php):

```suggestion
 {
    "packagistUrl": "https://packagist.org/packages/ecodev/graphql-upload",
    "keywords": ["graphql", "upload", "middleware"],
    "homepage": ""
  },
]
```

PowerKiKi commented on data/integration/integration-packages.json:
```diff
@@ -0,0 +1,172 @@
+    "packagistUrl": "https://packagist.org/packages/mimmi20/mezzio-setlocale-middleware",
+    "keywords": ["mezzio", "middleware", "language"],
+    "homepage": ""
+  }
```
I'd like to add one of my package too. It is a middleware to support file upload in GraphQL via [webonyx/graphql-php](https://github.com/webonyx/graphql-php):

```suggestion
  },
 {
    "packagistUrl": "https://packagist.org/packages/ecodev/graphql-upload",
    "keywords": ["graphql", "upload", "middleware"],
    "homepage": ""
  }
```

froschdesign commented:
@Jurj-Bogdan 
> As for the pagination, the partial is currently common between the integrations and blog lists - would "1, 2, 3 ..." buttons be desired in both places with the new design?

Can be done later.

arhimede commented:
preview updated, added @PowerKiKi package 

https://preview-1-hy2vwsq-2ja7ciew2nbkm.us-2.platformsh.site/

@froschdesign can we go live with it now ?

froschdesign commented:
@Jurj-Bogdan 
If there is a long package name, the image will be squeezed:

<img width="2374" height="1342" alt="laminas-integrations" src="https://github.com/user-attachments/assets/0aa85d7b-fe1f-4194-85d4-231fb6d2da04" />

Jurj-Bogdan commented:
Increased the page size to 24 and limited the max image size. Also, based on some feedback, changed the "Support Laminas" button on the home page to be more legible - seemed like too small a change for a separate PR

Jurj-Bogdan commented:
Added a json schema for easier validation of the integrations json, as suggested by @PowerKiKi, and updated the validation action to use it via `cardinalby/schema-validator-action` as discussed with @arhimede

arhimede commented:
@froschdesign  now we can go live ?

froschdesign commented:
@arhimede 
Please do not forget to promote this!
